### ISSUE TITLE
The ruler line falls behind the cursor

### ISSUE THREAD

**Issue by jdw1996** [tony612/highlight-column#32]
While using this package, I noticed that I was having the same problem as I was with the similar [`rulerz`](https://github.com/forgecrafted/rulerz) package. I described the problem in [an issue](https://github.com/forgecrafted/rulerz/issues/19) in the `rulerz` repository:

> When I type the first character on a new line, the ruler moves with the cursor no problem. When I type the second character on a line, the ruler doesn't move, so that it is then one column to the left of the cursor. If I continue typing, the ruler remains one column to the left of the cursor. I get the same problem when I start typing on a line that already has some characters on it.

I've noticed this problem on both Ubuntu 16.04 and Windows 7, using Atom 1.13.1.

**Comment by dodekeract** [comment]
Sounds like a CSS error in your theme/Stylesheet.

**Comment by jdw1996** [comment]
I commented out my whole stylesheet and tried it again and the same thing happened. Then I commented out my whole config file, same thing.

I get the same problem with every syntax theme I have installed, including all the defaults.

Sometimes (it seems to be when I hit two keys at almost the same time) the cursor will actually get to be two characters ahead of the line, as you can see in this picture:
![cursor_two_ahead](https://cloud.githubusercontent.com/assets/17225098/22577443/bc319cdc-e98f-11e6-8bda-1665f42c1ec9.png)

Here's a GIF showing the problem:
![highlight_line_malfunction](https://cloud.githubusercontent.com/assets/17225098/22577764/db4c4bce-e991-11e6-9522-c6809a3c5676.gif)

**Comment by dodekeract** [comment]
@jwd1996 Okay, if this only *sometimes* happens, especially depending on the typing speed it doesn't sound like CSS at all. That'd consistently happen.

**Comment by jdw1996** [comment]
As you can see in my GIF, the ruler is in the right position after you type your first character; that's always how it happens for me. Also, the ruler is always in the right spot when I'm backspacing.

I just noticed today that the line gets farther behind when I type faster.

**Comment by dodekeract** [comment]
@jwd1996 @tony612 I can actually reproduce this when typing **really** fast. (aka. smashing my keyboard)

Might be related to the CPU speed, using an i7-6700K I never noticed it in a realistic/programming setting.

@jwd1996 Any chance you're running this on a notebook/weak desktop?

**Comment by jdw1996** [comment]
@dodekeract I'm not sure at the moment what the specs are for my work computer (the one running Windows 7) but right now I'm seeing it on my laptop (the one running Ubuntu 16.04), which has an: Intel® Core™ i7-4510U CPU @ 2.00GHz × 4. I don't know much about hardware; let me know if there's any other information I can give you that might help.

**Comment by dodekeract** [comment]
@jwd1996 No, that's exactly what I'd have expected. So it seems that the code-base doesn't account for "too slow" execution of code/events. I'll take a look at it, but I think @tony612 probably already has an idea why this happens.

**Comment by jdw1996** [comment]
I don't know if this helpful anymore, but the processor on my work computer is an: Intel Core i7-4790 CPU @ 3.00 GHz 3.00 GHz.

**Comment by nathamanath** [comment]
I am experiencing a similar issue. Given the following code, first line is line 1, and I have soft tabs of 2 spaces:

```ex
defmodule FileAdapter do
  def parse(path) do
_
  end
end
```

When cursor is on row 3 column 1 (the underscore), and i press tab, cursor moves to column 5, however highlighted column moves to column 3.

**Comment by jdw1996** [comment]
@nathamanath Based on a quick search, it seems like this is Elixir code. Is that right?

I wasn't sure, but I tried it in a plain text file, and couldn't replicate, but that's at least partly because it can't recognize that extra indentation is needed in the third line without language support.

**Comment by nathamanath** [comment]
Hi @jdw1996, yes it is elixir, sorry I didn’t make that clear.
Having poked around a bit, I am getting this with other languages too.

I disabled all non core packages and added a breakpoint in the highlight column package, and took some screen shots to demonstrate what I am seeing:

https://transfer.sh/fBbNI/1.png
then I press return
https://transfer.sh/MDGGO/2.png
after continuing 
https://transfer.sh/N4L1A/3.png
then i press space
https://transfer.sh/13poz0/4.png
highlighted column is in line with cursor as expected

It seems the highlighted column is positioned before any auto indenting happens

**Comment by tony612** [comment]
The problem is `@editorElement.pixelRectForScreenRange`(in ./lib/highlight-column-element.coffee#L85) doesn't work as before. It's not a public API, so we have to find a solution to get pixel position of cursor.

**Comment by jdw1996** [comment]
Maybe [this comment](https://github.com/atom/atom/issues/13541#issuecomment-272032035) could be helpful? I tried substituting the code they suggest and it didn't work for me, but maybe you'll have more luck.

**Comment by tony612** [comment]
@jdw1996 Thank you. But it doesn't work. It behaves just like `pixelRectForScreenRange`.
